pkg/chunk/purger/purger.go

```diff
@@ -0,0 +1,576 @@
+	requestsWithBuildingPlanStatus, err := dp.deleteStore.GetDeleteRequestsByStatus(context.Background(), chunk.BuildingPlan)
+	if err != nil {
+		return err
+	}
+
+	for _, deleteRequest := range requestsWithBuildingPlanStatus {
+		req := makeDeleteRequestWithLogger(deleteRequest, util.Logger)
+
+		level.Info(req.logger).Log("msg", "loaded in process delete requests with status building plan")
+
+		dp.inProcessRequestIDs[deleteRequest.UserID] = deleteRequest.RequestID
+		err := dp.buildDeletePlan(req)
+		if err != nil {
+			level.Error(req.logger).Log("msg", "error building delete plan", "err", err)
+		}
+
+		level.Info(req.logger).Log("msg", "sending delete request for execution")
+		dp.executePlansChan <- req
+	}
+
+	requestsWithDeletingStatus, err := dp.deleteStore.GetDeleteRequestsByStatus(context.Background(), chunk.Deleting)
+	if err != nil {
+		return err
+	}
+
+	for _, deleteRequest := range requestsWithDeletingStatus {
+		req := makeDeleteRequestWithLogger(deleteRequest, util.Logger)
+		level.Info(req.logger).Log("msg", "loaded in process delete requests with status deleting")
+
+		dp.inProcessRequestIDs[deleteRequest.UserID] = deleteRequest.RequestID
+		dp.executePlansChan <- req
+	}
+
+	return nil
+}
+
+// pullDeleteRequestsToPlanDeletes pulls delete requests which do not have their delete plans built yet and sends them for building delete plans
+// after pulling delete requests for building plans, it updates its status to BuildingPlan status to avoid picking this up again next time
+func (dp *DataPurger) pullDeleteRequestsToPlanDeletes() error {
+	deleteRequests, err := dp.deleteStore.GetDeleteRequestsByStatus(context.Background(), chunk.Received)
+	if err != nil {
+		return err
+	}
+
+	for _, deleteRequest := range deleteRequests {
+		dp.inProcessRequestIDsMtx.RLock()
+		inprocessDeleteRequstID := dp.inProcessRequestIDs[deleteRequest.UserID]
+		dp.inProcessRequestIDsMtx.RUnlock()
+
+		if inprocessDeleteRequstID != "" {
+			level.Debug(util.Logger).Log("msg", "skipping delete request processing for now since another request from same user is already in process",
+				"inprocess_request_id", inprocessDeleteRequstID,
+				"skipped_request_id", deleteRequest.RequestID, "user_id", deleteRequest.UserID)
+			continue
+		}
+
+		if deleteRequest.CreatedAt.Add(24 * time.Hour).After(model.Now()) {
+			continue
+		}
+
+		err = dp.deleteStore.UpdateStatus(context.Background(), deleteRequest.UserID, deleteRequest.RequestID, chunk.BuildingPlan)
+		if err != nil {
+			return err
+		}
+
+		dp.inProcessRequestIDsMtx.Lock()
+		dp.inProcessRequestIDs[deleteRequest.UserID] = deleteRequest.RequestID
+		dp.inProcessRequestIDsMtx.Unlock()
+
+		req := makeDeleteRequestWithLogger(deleteRequest, util.Logger)
+
+		level.Info(req.logger).Log("msg", "building plan for a new delete request")
+
+		err := dp.buildDeletePlan(req)
+		if err != nil {
+			level.Error(req.logger).Log("msg", "error building delete plan", "err", err)
```

Comment:
`inProcessRequestIDs` is not cleared here, so no more deleting for user will happen.

Reply:
It is cleared [here](https://github.com/cortexproject/cortex/blob/2819c3d18af56d68878cb15d4c120248acb2b656/pkg/chunk/purger/purger.go#L173), when the delete request is marked `Processed`. We need to make sure there is only 1 delete requests in `buildingPlan` or `deleting` state to avoid overlapping delete requests causing problems.

Reply:
I don't think control can get to that point from here. Error would need to be passed via `workerJobExecutionStatusChan`, but I this method is not run by worker at all.

Reply:
Ohh right. I got your question now. Actually I deliberately kept it this way so that we fix the problem of failed delete request first and avoid moving multiple delete requests in the deletion process. Will add a comment for future readers.